04week/mastermind.js

```diff
@@ -28,15 +31,69 @@ function getRandomInt(min, max) {
   return Math.floor(Math.random() * (max - min)) + min;
 }
 
-function generateHint() {
-  // your code here
+function generateHint(guess) {
+  let guessArray = guess.toUpperCase().trim().split('');
+  let solutionArray = solution.toUpperCase().split('');
+  let correctLetterLocations= 0;
+  let correctLetters = 0;
+  let turnCount = board.length+1;
+
+// find correct letters that are also in the correct location and store them in correctLetterLocations
+  for (let i=0; i < solutionArray.length; i++) {
+    if (solutionArray[i] == guessArray[i]) {
+      correctLetterLocations++
+      solutionArray[i] = null
+      //console.log("correct letter and location") 
+    } //else { //lines 46-48 used for testing
+      //console.log("no letter + location match found")
+    //}
+  }
+
+// find correct letters that are NOT in the correct location and store them in correctLetters
+  for (let i=0; i < solutionArray.length; i++) {
+    let targetIndex = solutionArray.indexOf(guessArray[i]); 
+    if (targetIndex > -1) {
+      correctLetters++
+      solutionArray[targetIndex] = null
+    } //else { //lines 57-59 used for testing
+      //console.log("no letter match found")
+    //}
+  };
+  // uncomment lines 62-65 for colorful fancy game play****
+  //hint = `
+  //****Correct Location: ${correctLetterLocations}`.red
+  //+`\n******Correct Letter: ${correctLetters}`.white 
+  //+`\n**********Your Guess: ${guess}`.blue
+
+  // comment out lines 68-69 for colorful fancy game play (neeeded to pass tests)
+  hint = `${correctLetterLocations}-${correctLetters}`; 
+  return hint; 
 }
 
 function mastermind(guess) {
-  solution = 'abcd'; // Comment this out to generate a random solution
-  // your code here
-}
+  //solution = 'abcd'; // Comment this out to generate a random solution
+  let turnCount = board.length+1
+  //check to see if the player has guessed the solution
+  //if the player won, they receive a notification, and the game board is reset for a new game to begin;
+  //if the player did not guess correctly a hint is generated
+  //the players incorrect guess + the hint are displayed on the "board"
+  if (guess == solution) {
+    console.log("You cracked the code!".america) // altered text will not pass the unit test
+    console.log("***Let's play again!***".rainbow)
+    board = [];
+    return `You guessed it!`; // this is needed to pass the unit test
+  } else if (board.length <=8 ) {
+    generateHint(guess);
```

Comment:
this method returns the hint. Instead of making hint global, it would be better to assign it here.